### PR TITLE
Add support for ReplicaSet k8s kind

### PR DESCRIPTION
**Pull request by anastasia-malysheva**
Signed-off-by: anastasia.malysheva <anastasia.malysheva@xored.com>

### CODE REVIEW

**Comment by anastasia-malysheva** [comment]
As part of the task it was found that Job and CronJob K8s resourses can be supported with nsm annotations too, but it is not necessary now so it was not added.

**Comment by anastasia-malysheva** [comment]
As part of the task the small research was made about usage of Annotations in service meshes, Istio and Consul-k8s. For Consul annotations are widely used for pods, but also they are used for Services and Service Accounts and there are examples with custom kinds. For Istio annotations used also for Pods, Deployments, Services and other types too.

File: main.go [hunk 1]
```diff
@@ -133,10 +152,37 @@ func (s *admissionWebhookServer) unmarshal(in *admissionv1.AdmissionRequest) (p
...
 	p = path.Join("/", p)
-	if metaPtr.Labels == nil {
-		metaPtr.Labels = make(map[string]string)
+	if podMetaPtr.Labels == nil {
+		podMetaPtr.Labels = make(map[string]string)
+	}
+	// Annotations shouldn't be applied second time.
+	if isReplicaOwnedByDeployment(in.Kind.Kind, metaPtr) {
+		return "", nil, nil
+	}
```
Comment (denis-tingaikin) [inline main.go:161]: @anastasia-malysheva Could you say more about this?

File: main.go [hunk 2]
```diff
@@ -133,10 +152,37 @@ func (s *admissionWebhookServer) unmarshal(in *admissionv1.AdmissionRequest) (p
...
+		for _, o := range metaPtr.OwnerReferences {
+			if o.Kind == deploymentKind {
+				return true
+			}
+		}
+	}
+	return false
+}
+
+func updatePodAnnotations(kind string, metaPtr, podMetaPtr *v1.ObjectMeta) {
```
Comment (denis-tingaikin) [inline main.go:178]: Why we cant do this directly in pod branch in the switch case above?

File: main.go [hunk 3]
```diff
@@ -133,10 +152,37 @@ func (s *admissionWebhookServer) unmarshal(in *admissionv1.AdmissionRequest) (p
...
+	if kind != podKind && metaPtr.Annotations != nil {
+		if podMetaPtr.Annotations == nil {
+			podMetaPtr.Annotations = metaPtr.Annotations
+		}
+		err := errors.New("can't register a sink factory for empty string")
+		panic(err.Error())
```
Comment (denis-tingaikin) [inline main.go:184]: Can we don't panic? It is not good practice.

**Review by denis-tingaikin — CHANGES_REQUESTED** [review]

File: main.go (hunk 1 above)
Comment (anastasia-malysheva) [reply main.go:161]: @denis-tingaikin, replicaSet inherits annotations from the deployment it is owned by. As soon as deployment already has annotations and they've been applied before replicaset was created, replicaset tries to apply same annotations again and it causes errors during installation. That's why this check is here.

File: main.go (hunk 3 above)
Comment (anastasia-malysheva) [reply main.go:184]: Yes, I'll fix it.

File: main.go (hunk 2 above)
Comment (anastasia-malysheva) [reply main.go:178]: Because the method is already big enough and lint raise errors when I add more code there.

Comment (denis-tingaikin) [reply main.go:178]: Can we fix the linter?

**Comment by denis-tingaikin** [comment]
I think this is good. But it requires a small refactor. I see two options:

1. make new methods inline and fix linter 
2. create for each type in switch operator a handle function. Something like 
```go
switch t {
case "pod":
handlePod(...)
...
}
```

File: main.go (hunk 3 above)
Comment (anastasia-malysheva) [reply main.go]: Replaced with logger error.

File: main.go (hunk 2 above)
Comment (anastasia-malysheva) [reply main.go]: I used inline function as you proposed and it worked, so there is no need to change linter settings.

**Comment by denis-tingaikin** [comment]
@anastasia-malysheva Could you please resolve these comments 
https://github.com/networkservicemesh/cmd-admission-webhook-k8s/pull/152#discussion_r918052401
https://github.com/networkservicemesh/cmd-admission-webhook-k8s/pull/152#discussion_r918052268

File: main.go [hunk 4]
```diff
@@ -100,43 +106,67 @@ func (s *admissionWebhookServer) Review(in *admissionv1.AdmissionRequest) *admis
...
+		podMetaPtr.Labels = make(map[string]string)
+	}
+	// Annotations shouldn't be applied second time.
+	if in.Kind.Kind == replicaSetKind {
+		for _, o := range metaPtr.OwnerReferences {
+			if o.Kind == deploymentKind {
+				return "", nil, nil
+			}
+		}
 	}
```
Comment (denis-tingaikin) [inline main.go:160]: Can we move this into replicaSetKind case at line 137?

File: main.go [hunk 5]
```diff
@@ -100,43 +106,67 @@ func (s *admissionWebhookServer) Review(in *admissionv1.AdmissionRequest) *admis
...
-	return p, metaPtr, podSpec
+
+	if in.Kind.Kind != podKind && metaPtr.Annotations != nil {
+		if podMetaPtr.Annotations == nil {
+			podMetaPtr.Annotations = metaPtr.Annotations
+		}
+		s.logger.Errorf("Malformed specification. Annotations can't be provided in several places.")
+	}
```
Comment (denis-tingaikin) [inline main.go:167]: Can we move this into pod case at line 119?

File: main.go (hunk 4 above)
Comment (anastasia-malysheva) [reply main.go:160]: No, we can't because unmarshal to target is performed after switch.

File: main.go (hunk 5 above)
Comment (denis-tingaikin) [reply main.go:167]: Could we please use a simple defer?

File: main.go (hunk 4 above)
Comment (denis-tingaikin) [reply main.go:160]: Could we please use a simple defer?

Comment (anastasia-malysheva) [reply main.go]: Fix was added.

File: main.go [hunk 6]
```diff
@@ -100,43 +106,68 @@ func (s *admissionWebhookServer) Review(in *admissionv1.AdmissionRequest) *admis
...
-	case "Deployment":
+	case deploymentKind:
 		var deployment appsv1.Deployment
-		metaPtr = &deployment.Spec.Template.ObjectMeta
+		metaPtr = &deployment.ObjectMeta
+		podMetaPtr = &deployment.Spec.Template.ObjectMeta
 		podSpec = &deployment.Spec.Template.Spec
 		target = &deployment
-	case "Pod":
+	case podKind:
```
Comment (denis-tingaikin) [inline main.go:119]:
```suggestion
	case "Pod":
```

File: main.go [hunk 7]
```diff
@@ -100,43 +106,68 @@ func (s *admissionWebhookServer) Review(in *admissionv1.AdmissionRequest) *admis
...
+		podSpec = &replicaSet.Spec.Template.Spec
+		target = &replicaSet
+		defer func() {
+			s.logger.Info("Replicaset Defer method")
+			for _, o := range metaPtr.OwnerReferences {
+				if o.Kind == deploymentKind {
+					p, meta, spec = "", nil, nil
+				}
+			}
+		}()
```
Comment (denis-tingaikin) [inline main.go:150]: This is still looking overcomplicated.

I think the main problem here that we are doing things in `unmarshal` method that actually doesnt related to the operation.

Please move all  'post unmarshal' changes into separate method and use it right afiter unmarshal in the Review function.

File: main.go [hunk 8]
```diff
@@ -56,6 +56,12 @@ type admissionWebhookServer struct {
 	logger *zap.SugaredLogger
 }
 
+const (
+	deploymentKind string = "Deployment"
+	podKind        string = "Pod"
+	replicaSetKind string = "ReplicaSet"
+)
+
```
[the remark this answers is not in the document]
Comment (denis-tingaikin) [reply main.go:64]: @anastasia-malysheva Could you please resolve this comment?

[the remark this answers is not in the document]
Comment (anastasia-malysheva) [reply main.go:64]: @denis-tingaikin I took a look to the k8s doc and couldn't find constants for this

[the remark this answers is not in the document]
Comment (anastasia-malysheva) [reply main.go]: @denis-tingaikin I changed linter setting and removed constants

File: main.go (hunk 7 above)
Comment (anastasia-malysheva) [reply main.go]: I've added changes, please take a look

**Review by denis-tingaikin — APPROVED** [review]

**Pull request merged**